nexusmods-app: refactor the unfree variant

MattSturgeon opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

- Reformat with nixfmt
- ~~Cherry picked some fixes by @corngood.
  If #327651 is merged before this, they can be dropped in a rebase.~~
- Removed the `enableUnfree` argument; instead users should override the `_7zz` argument.
- Refactor **nexusmods-app-unfree**'s implementation to use `override` instead of `callPackage`
- Set **nexusmods-app-unfree**'s `pname` using `overrideAttrs`
- Add a long description, which varies depending on `_7zz.meta.unfree`
  - This attempts to warn users that the free variant doesn't support RAR archive mods
  - It also documents how to enable RAR support

Merging this early should hopefully make #331150 easier to work on and review.

See also: #327651 #331150
Closes #321405

cc @l0b0 @corngood @GGG-KILLER @erri120

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

MattSturgeon committed:
nexusmods-app: reformat

- Format using nixfmt
- Apply other minor formatting tweaks
- Simplify `--filter` test flag definition

MattSturgeon committed:
nexusmods-app: add long description

Add a long description that documents whether or not RAR archive mods
are supported and how to enable support.

MattSturgeon committed:
nexusmods-app: refactor "unfree" variant

Use `override` to pass a `_7zz` with RAR support to the package, instead
of having an `enableUnfree` package argument.

Use a different `pname` on the unfree package.

Use `_7zz.meta.unfree` internally.

MattSturgeon commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -63,25 +64,28 @@ buildDotnetModule rec {
-  executables = [
-    nexusmods-app.meta.mainProgram
-  ];
+  executables = [ nexusmods-app.meta.mainProgram ];
```
@SuperSandro2000 do you still want references to `meta.mainProgram` removed? Should I tackle that in this PR?

corngood replied:
My suggestion would be to just `let executable = ...` at the top.  `nexusmods-app` is not necessarily the final package we're evaluating here.

There's a similar problem with tests, where `nexusmods-app.tests` and `nexusmods-app-unfree.tests` evaluate to the same thing.

I think ideally we'd avoid referencing ourself in the `callPackage` args, but we probably need `buildDotnetModule` to support `finalAttrs` first.

MattSturgeon replied:
> My suggestion would be to just `let executable = ...` at the top. `nexusmods-app` is not necessarily the final package we're evaluating here.

Since we're already using a `rec` attrset, we could directly reference `meta.mainProgram` too, but that won't solve the test cases.

> I think ideally we'd avoid referencing ourself in the `callPackage` args, but we probably need `buildDotnetModule` to support `finalAttrs` first.

Agreed, I think that is needed to build the tests properly.

I'll take a quick look at adding support to `buildDornetModule`.

corngood replied:
The `rec` thing sounds good to me too, and we don't have to fix the tests in this PR if it turns into a big project.

MattSturgeon commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -108,7 +112,6 @@ buildDotnetModule rec {
   };
```
How should the tests reference the package's exe without explicitly referencing `pkgs.nexusmods-app`?

Presumably this may use the wrong derivation when we're building an overridden package?

MattSturgeon replied:
Seems most other packages just set `nativeBuildInputs = [ nexusmods-app ]` in the test derivations.

Maybe the `-unfree` override in all-packages.nix should also be overriding the `nexusmods-app` argument passed in?

MattSturgeon replied:
I'm going to look at [these examples](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20finalAttrs.finalPackage&type=code) using the `finalAttrs` pattern tomorrow if I get chance

corngood replied:
Sorry, I responded to the other thread pretty much saying what you said here.

corngood commented:
Sorry for the conflict, but it's trivial and you should be able to rebase now to drop the change you cherry-picked.

MattSturgeon committed:
nexusmods-app: cleanup tests & `mainProgram` refs

We still need to fix the tests to use `finalAttrs`, but that requires
changes to `buildDotnetModule`.

corngood commented:
> Removed the enableUnfree argument; instead users should override the _7zz argument.

What's the rationale for this?  It feels like it leaks a bunch of details about what makes the package free vs. unfree into `all-packages.nix`.  `enableUnfree` seems to be used in a few places already.

MattSturgeon commented:
> > Removed the `enableUnfree` argument; instead users should override the `_7zz` argument.
> 
> What's the rationale for this? It feels like it leaks a bunch of details about what makes the package free vs. unfree into `all-packages.nix`. `enableUnfree` seems to be used in a few other places.

My thought process was that it's not really the `nexusmods-app` that is free/unfree, but rather its dependency.

I haven't looked too hard, but I assume `enableUnfree` is normally used when the package itself has an optional unfree feature or multiple optional unfree dependencies?

My assumption was that it'd be better to remove the abstraction, exposing the details to end-users who wanted to override the package themselves.

---

I'm happy to add it back, since it is probably a simpler interface for end-users. Although saying that, `nexusmods-app-unfree` itself is _intended_ to be the "simple" interface.

While I've been using the nix language and module systems for a bit, this is my first time _packaging_ something, so I appreciate all the feedback and guidance I can get!

If we add the `enableUnfree` argument back, should it be done in such a way that users can still provide an unfree `_7zz` without us then overriding its `enableUnfree` argument?

e.g. `pkgs.nexusmods-app.override { _7zz = pkgs._7zz.override { enableUnfree = true; }; }` wouldn't work without this PR, because `nexusmods-app`'s `enableUnfree` argument would internally override `_7zz`'s back to false.

corngood commented:
What if we did:

```
  nexusmods-app-unfree = callPackage ../by-name/ne/nexusmods-app/package.nix {
    pname = "nexusmods-app-unfree";
    _7zz = _7zz.override { enableUnfree = true; };
  };
```

and add `pname ? "nexusmods-app"` to the package?

I don't really have a strong opinion about this.  There are various examples that support any of these methods:

```
  openocd-rp2040 = openocd.overrideAttrs (old: {
    pname = "openocd-rp2040";
    src = fetchFromGitHub {
      owner = "raspberrypi";
      repo = "openocd";
      rev = "4d87f6dcae77d3cbcd8ac3f7dc887adf46ffa504";
      hash = "sha256-bBqVoHsnNoaC2t8hqcduI8GGlO0VDMUovCB0HC+rxvc=";
      # openocd disables the vendored libraries that use submodules and replaces them with nix versions.
      # this works out as one of the submodule sources seems to be flakey.
      fetchSubmodules = false;
    };
    nativeBuildInputs = old.nativeBuildInputs ++ [
      autoreconfHook
    ];
  });
```

```
  firefox-bin = wrapFirefox firefox-bin-unwrapped {
    pname = "firefox-bin";
  };
```

A couple of other minor things:

- I think we can drop `pkgs.` since there's a `with pkgs;` at the top
- Should we use `_7zz-rar` instead of doing the override?  If rar support is what we're looking for, then I think it matches the intent better.  Imagine there's an alternate implementation of 7zz with rar support in the future.

MattSturgeon commented:
> What if we [used `callPackage`] and add `pname ? "nexusmods-app"` to the package?

I considered adding a `pname` argument to the package but wasn't sure if that was the "done thing" instead of using `overrideAttrs`. I agree that chaining both `override` and `overrideAttrs` is kinda messy though!

Is it better to use `callPackage` or `override` for package variants? I'd assumed the latter was preferred.

> I think we can drop `pkgs.` since there's a `with pkgs;` at the top

Thanks, I didn't spot that!

> Should we use `_7zz-rar` instead of doing the override?  If rar support is what we're looking for, then I think it matches the intent better.  Imagine there's an alternate implementation of 7zz with rar support in the future.

Yes, I think so. I don't think either me or @l0b0 spotted that package when initially adding nexusmods-app.

Should `nexusmods-app-unfree` be renamed `nexusmods-app-rar` in line with that? We probably wouldn't need an alias for very long given how new the package is.

MattSturgeon commented:
Looking at `_7zz` and `_7zz-rar`, both have the same pname; maybe it isn't important to change the pname for the nexusmods-app variant that has RAR support?

https://github.com/nixos/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L1466-L1467

corngood commented:
> Should nexusmods-app-unfree be renamed nexusmods-app-rar in line with that? We probably wouldn't need an alias for very long given how new the package is.

I think probably not.  It only really exists as a variant because there's no free way to support rar.  In the event that that changes, we'd probably just deprecate -unfree.

> I considered adding a pname argument to the package but wasn't sure if that was the "done thing" instead of using overrideAttrs

One minor argument against `overrideAttrs` is that if the package ever switched to use `name` instead of `pname`, the override would have no effect.  Having it in the package args is a hint that it's overridden.  There are a few examples around if you search for `pname ? "`, e.g. `pname ? "gnuradio"`.

> maybe it isn't important to change the pname for the nexusmods-app variant that has RAR support?

I think this could also be fine.

> Is it better to use callPackage or override for package variants? I'd assumed the latter was preferred.

I would have actually thought the former, because of how splicing works in `callPackage`.  Don't take my word for it though, because I don't fully understand all of the details.  Also, the differences are probably not a practical concern for this package.

There's so much variety in all-packages that realistically any of these options are probably okay.

MattSturgeon commented:
@corngood thanks for your prompt reviews and helpful feedback!

l0b0 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -1,26 +1,21 @@
+  libX11,
+  nexusmods-app,
+  runCommand,
+  pname ? "nexusmods-app",
```
Why would I want to optionally change the package name to an arbitrary string, when in reality there's only two useful package names? I'm pretty sure the original pattern of using `enableUnfree` (or something like it) is how it's done elsewhere. Having to override `_7zz` is much more technically difficult than setting `enableUnfree = true`. And finally, `enableUnfree` means we can easily add more optional unfree components if necessary without changing the function signature.

MattSturgeon replied:
Regarding `pname` as a package argument: that was discussed in a few comments above, mostly from here onwards: https://github.com/NixOS/nixpkgs/pull/331355#issuecomment-2262888714

Regarding the motivation for dropping `enableUnfree`, see this comment: https://github.com/NixOS/nixpkgs/pull/331355#issuecomment-2262848081

> Having to override _7zz is much more technically difficult than setting enableUnfree = true.

I don't agree that `_7zz = pkgs _7zz-rar` is **significantly** more difficult than `enableUnfree = true`.

I also prefer how the former is intrinsically clear about its purpose.

At first glance, the later could be interpreted as any number of unfree things being enabled.

> And finally, enableUnfree means we can easily add more optional unfree components if necessary without changing the function signature.

Maybe I'll eat my words, but I don't see that as a likely scenario. Nexusmods-app is committed to being an open source project with open source dependencies.

It is just unfortunate that one  optional feature of one dependency currently requires an unfree license.

If there were multiple unfree features to be enabled, then an `enableUnfree` argument would be more useful, I agree. If that happens we can revisit this 😀

l0b0 replied:
> I also prefer how the former is intrinsically clear about its purpose.

I prefer how the latter is more clear about the thing I care about. But of course people care about different things. :grin:

I still prefer the original design, but I'll let other reviewers decide.

l0b0 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -1,26 +1,21 @@
-, runCommand
-, enableUnfree ? false # Set to true to support RAR format mods
+{
+  lib,
```
Did `nixfmt` change this ordering? It seems weird having only `lib` on top.

MattSturgeon replied:
No, that was manual.

I sorted almost all args, but decided to keep `lib` and our "custom" args (pname) unsorted as (in my head) they are distinct.

I can sort all args or all auto-args if you prefer, but is seems like a minor point.

l0b0 replied:
I agree `pname` should be last, which is why I didn't mention it. It is a minor point, but having to read through a big list of parameters when I could just scan it alphabetically is tedious.

MattSturgeon replied:
I'll fix this tomorrow, heading to sleep now.

Thanks for your feedback!

MattSturgeon replied:
I've reverted to everything other than `pname` being sorted alphanumerically

l0b0 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -72,24 +72,29 @@ buildDotnetModule rec {
-      )
-    )
-  ];
+        ++ lib.map (n: "FullyQualifiedName!=${n}") (
```
This seems like premature abstraction (as soon as something other than `FullyQualifiedName!=${n}` is necessary for `!enableUnfree` this would have to be unpicked), but feel free to leave it in if you prefer.

MattSturgeon replied:
This is basically just re-implementing `disabledTests` from buildDotnetModule but with support for having additional filters.

I think it's unlikely we'll need other filters optionally, but if we do we can either have multiple conditions or deal with it some other way at the time.

The best (long-term)
solution is that buildDotnetModule adds support for additional `testFilters`, so we don't have to build the `--filter` flag ourselves.

l0b0 replied:
I think we're talking about different things, but feel free to leave it in.

MattSturgeon replied:
I've dropped these changes

l0b0 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -58,7 +58,7 @@ buildDotnetModule rec {
   makeWrapperArgs = [
     "--prefix PATH : ${lib.makeBinPath [ desktop-file-utils ]}"
-    "--set APPIMAGE $out/bin/${meta.mainProgram}" # Make associating with nxm links work on Linux
+    "--set APPIMAGE ${placeholder "out"}/bin/${meta.mainProgram}" # Make associating with nxm links work on Linux
```
What's the advantage of `${placeholder "out"}` over `$out`?

MattSturgeon replied:
Just using `"$out"` didn't work correctly; it ends up being included literally in the wrapper.

This was taken from feedback on one of the other PRs.

l0b0 replied:
:open_mouth: OK then, my bad! It (that is, how this was done *some time ago*) *used* to work, but I must've messed up.

corngood commented:
> Is that a packaging issue on my end, or do the tests always queue for all systems?

I'm not 100% sure, but my guess is it's not fully checking metadata before queueing and will just end up doing nothing.

l0b0 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
````diff
@@ -136,5 +135,29 @@ buildDotnetModule rec {
+            ```
+          ''
+      }
+    '';
````
The descriptions should probably be at or near the top of the `meta`. Most packages seem to have this crucial information near the top A quick scan seems to confirm this.

MattSturgeon replied:
I wanted `longDescription` near the bottom because it is, well, kinda long. So I ended up moving both descriptions.

SuperSandro2000 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -127,5 +130,29 @@ buildDotnetModule rec {
     platforms = lib.platforms.linux;
+    description = "Game mod installer, creator and manager";
+    longDescription = ''
+      ${meta.description}.
+
```
```suggestion
```
longDescription should not just repeat description as they are often rendered next to each other

MattSturgeon replied:
Are they _always_ rendered together? Would including the more verbose description from upstream be ok?

MattSturgeon replied:
[Upstream's readme](https://github.com/Nexus-Mods/NexusMods.App/blob/main/README.md) is as follows:

> Nexus Mods app is a mod installer, creator and manager for all your popular games.
> 
> Easy to use, runs on your standard Windows PC and Linux alike. Don't waste time troubleshooting, play your games,
> fill those knees with arrows and most importantly, ***Have Fun***!
> 
> Learn more about the App on the [Wiki](https://nexus-mods.github.io/NexusMods.App/)

Perhaps the long description could include some of that along with a note that the app is still experimental?

e.g.:

```suggestion
      A mod installer, creator and manager for all your popular games.

      Currently experimental and undergoing active development,
      new releases may include breaking changes!

```

SuperSandro2000 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
````diff
@@ -127,5 +130,29 @@ buildDotnetModule rec {
+          ''
+            It is strongly recommended that you use the "unfree" variant of this package,
+            which provides support for mods packaged as RAR archives.
+
+            You can also enable unrar support manually, by overriding the `_7zz` used:
+
+            ```nix
+            pkgs.nexusmods-app.override {
+              _7zz = pkgs._7zz-rar;
+            }
+            ```
````
```suggestion
```
What's the point off this?

MattSturgeon replied:
More documentation is usually better, right? It sheds some light on what the "unfree" version is actually doing and empowers users to customize the package to their liking.

If you still think this is too verbose for the `longDescription`, at least we still have the paragraph above regarding `nexusmods-app-unfree`.

SuperSandro2000 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -72,52 +65,62 @@ buildDotnetModule rec {
-  executables = [
-    nexusmods-app.meta.mainProgram
-  ];
+  executables = [ meta.mainProgram ];
```
```suggestion
  executables = [ "NexusMods.App" ];
```

MattSturgeon replied:
Thanks for the feedback, but could you clarify the motivation?

When I discussed this earlier (https://github.com/NixOS/nixpkgs/pull/331355#discussion_r1699005403) it was thought that using `meta.mainProgram` via `rec` or `finalAttrs` would be fine.

corngood replied:
@SuperSandro2000 

You asked for the same thing in https://github.com/NixOS/nixpkgs/pull/331150#discussion_r1698368777, but never explained your reasoning.

SuperSandro2000 commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -54,15 +47,15 @@ buildDotnetModule rec {
-    "--prefix PATH : ${lib.makeBinPath [desktop-file-utils]}"
-    "--set APPIMAGE $out/bin/${meta.mainProgram}" # Make associating with nxm links work on Linux
+    "--prefix PATH : ${lib.makeBinPath [ desktop-file-utils ]}"
+    "--set APPIMAGE ${placeholder "out"}/bin/${meta.mainProgram}" # Make associating with nxm links work on Linux
```
```suggestion
    "--set APPIMAGE ${placeholder "out"}/bin/NexusMods.App" # Make associating with nxm links work on Linux
```

MattSturgeon commented:
> > Is that a packaging issue on my end, or do the tests always queue for all systems?
> 
> I'm not 100% sure, but my guess is it's not fully checking metadata before queueing and will just end up doing nothing.

Maybe we can use `meta.hydraPlatforms` for this :thinking:

MattSturgeon commented:
I'm conscious that the nexus team plan to have another release out in a week or so, it'd be nice to get cleanup and refactoring out of the way so that we can refocus on getting the package updated and working.

Other than taking advantage of `finalAttrs` attr fixing that `buildDotnetModule` recently gained support for, how would you recommend we go about addressing the feedback so that we can move forward?

Should I adapt this PR to use `finalAttrs`-style arguments? I guess that may resolve _some_ of @SuperSandro2000's feedback 🤔

@corngood @l0b0 @SuperSandro2000

corngood commented:
You've posted reasonable responses to all the comments and there have been no responses, so I'm happy to merge this when you think it's ready.

> Should I adapt this PR to use finalAttrs-style arguments?

Your call.  I don't mind if it's done separately.  `omnisharp-roslyn` also needs a fix.

MattSturgeon commented:
> You've posted reasonable responses to all the comments and there have been no responses, so I'm happy to merge this when you think it's ready.
> 
> > Should I adapt this PR to use finalAttrs-style arguments?
> 
> Your call. I don't mind if it's done separately. `omnisharp-roslyn` also needs a fix.

I've applied the changes from https://github.com/NixOS/nixpkgs/pull/331355#discussion_r1705509681, but otherwise I think the feedback is either personal preferences, or best addressed while fixing the recursive definitions to use `finalAttrs`.

I think that's best done in a fresh PR, so if you're happy, I'm happy. Let's get this merged!

corngood commented:
Result of `nixpkgs-review pr 331355` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>nexusmods-app</li>
    <li>nexusmods-app-unfree</li>
  </ul>
</details>
